src/com/sun/jna/internal/Cleaner.java

```diff
@@ -35,154 +39,226 @@
+                        try {
+                            Reference<?> ref = impl.referenceQueue.remove(MASTER_CLEANUP_INTERVAL_MS);
+                            if(ref instanceof CleanerRef) {
+                                ((CleanerRef) ref).clean();
+                            }
+                            // "now" is not really *now* at this point, but off by no more than MASTER_CLEANUP_INTERVAL_MS
+                            if (lastMasterRun + MASTER_CLEANUP_INTERVAL_MS <= now) {
+                                masterCleanup();
+                                lastMasterRun = now;
+                            }
+                        } catch (InterruptedException ex) {
+                            // Can be raised on shutdown. If anyone else messes with
+                            // our reference queue, well, there is no way to separate
+                            // the two cases.
+                            // https://groups.google.com/g/jna-users/c/j0fw96PlOpM/m/vbwNIb2pBQAJ
+                            break;
+                        } catch (Exception ex) {
+                            Logger.getLogger(Cleaner.class.getName()).log(Level.SEVERE, null, ex);
+                        }
+                    }
+                    LOG.log(Level.FINE, "MasterCleaner thread {0} exiting", Thread.currentThread());
+                }, "JNA Cleaner");
+            LOG.log(Level.FINE, "Starting new MasterCleaner thread {0}", cleanerThread);
+            cleanerThread.setDaemon(true);
+            cleanerThread.start();
         }
 
-        CleanerRef getNext() {
-            return next;
+        private void masterCleanup() {
+            for (Iterator<Map.Entry<Thread,Cleaner>> it = Cleaner.INSTANCES.entrySet().iterator(); it.hasNext(); ) {
+                Map.Entry<Thread,Cleaner> entry = it.next();
+                if (!cleaners.contains(entry.getValue())) { continue; }
+                Cleaner cleaner = entry.getValue();
+                long currentCount = cleaner.counter.get();
+                if (currentCount == cleaner.lastCount // no new cleanables registered since last master cleanup interval -> assume it is no longer in use
+                        || !entry.getKey().isAlive()) { // owning thread died -> assume it is no longer in use
+                    it.remove();
+                    CleanerImpl impl = cleaner.impl;
+                    LOG.log(Level.FINE, () -> "MasterCleaner stealing cleaner " + impl + " from thread " + entry.getKey());
+                    referencedCleaners.add(impl);
+                    watchedCleaners.add(impl);
+                    register(cleaner, () -> {
+                        referencedCleaners.remove(impl);
+                        LOG.log(Level.FINE, "Cleaner {0} no longer referenced", impl);
+                    });
+                    cleaners.remove(cleaner);
+                } else {
+                    cleaner.lastCount = currentCount;
+                }
```

Comment:
> Consider a thread that does native work, then sleeps for a time longer time than MASTER_CLEANUP_INTERVAL_MS, and does this in a loop. While doing this a native reference is kept in the Thread. Then in each interation a new CleanerImpl will be constructed.
> ...
> The former one needs adoption to handle the possible race where the check for "no cleanables" happens

Yes.

I did consider what you suggest, and decided against it because I couldn't see a way to avoid the race without introducing some kind of synchronization in the Cleaner.. The original Cleaner implementation seemed to deliberately avoid synchronization, so I wanted to avoid it too. The price to be paid for this is the possibility of having more Cleaner instances than strictly required, but I think the overhead is small except in rather degenerate cases.